exercises/lamar_moore/ex02_oo_basics/src/ShapeSorter.cpp

```diff
@@ -0,0 +1,172 @@
+#include <iostream>
+#include <algorithm>
+#include <memory>
+#include "ShapeSorter.h"
+
+
+using namespace std;
+
+namespace Shapes
+{
+	/** Functor which is used to compare shape types with a matching value in std::remove_if
+	*/
+	class MatchTypeFunctor
+	{
+		ShapeType type;
+	public:
+		MatchTypeFunctor(ShapeType t): type(t)
+		{
+		}
+
+		bool operator()(shared_ptr<Shape> shape) const 
+		{
+			bool isMatch = true;
+
+			if(shape->getType() == type)
+				isMatch = false;
+
+			return isMatch;
+		}
+	};
+
+	/** Functor which is used to compare shape types with a matching value in std::remove_if
+	*/
+	class MatchSideFunctor
+	{
+		int numSides;
+	public:
+		MatchSideFunctor(int sides): numSides(sides)
+		{
+		}
+
+		bool operator()(shared_ptr<Shape> shape) const 
+		{
+			bool isMatch = true;
+
+			if(shape->getNumSides() == numSides)
+				isMatch = false;
+
+			return isMatch;
+		}
+	};
+
+	/** Functor which compares the magnitude of the areas between two Shapes
+	*/
+	class AreaFunctor
+	{
+	public:
+		bool operator() (shared_ptr<Shape> first, shared_ptr<Shape> second)
+		{
+			return first->calculateArea() < second->calculateArea();
+		}
+	};
+
+	/** Functor which compares the magnitude of the perimeter between two shapes
+	*/
+	class PerimeterFunctor
+	{
+	public:
+		bool operator() (shared_ptr<Shape> first, shared_ptr<Shape> second)
+		{
+			return first->calculatePerimeter() < second->calculatePerimeter();
+		}
+	};
+
+	/** Constructor
+	*/
+	ShapeSorter::ShapeSorter()
+	{
+	}
+
+	/** Destructor
+	*/
+	ShapeSorter::~ShapeSorter()
+	{
+	}
+
+	/** Prints all shapes which have the matching type specified by the first argument
+
+	@param type Shape type for comparison with list of shapes
+	@param shapes List of shapes for which types will be selectively printed
+	*/
+	void ShapeSorter::printByType(ShapeType type, vector<shared_ptr<Shape>> &shapes)
+	{
+		vector<shared_ptr<Shape>> copy = shapes;
```

Comment:
I think you could save a lot of allocations deallocations if you used std::copy_if. But I didn't think about that until I looked at the stl.